In [1]:
import numpy as np
import pandas as pd

In [2]:
df_user = pd.read_csv('E:/Data/BX-Users.csv',encoding='latin-1')

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3146: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [3]:
df_user.head()

user_id                            Location   Age
0       1                  nyc, new york, usa   NaN
1       2           stockton, california, usa  18.0
2       3     moscow, yukon territory, russia   NaN
3       4           porto, v.n.gaia, portugal  17.0
4       5  farnborough, hants, united kingdom   NaN

In [5]:
#column_names = ['isbn', 'book_title']
df_books = pd.read_csv('E:/Data/BX-Books.csv', encoding='latin-1')

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3146: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [6]:
df_books.head()

isbn                                         book_title  \
0  195153448                                Classical Mythology   
1    2005018                                       Clara Callan   
2   60973129                               Decision in Normandy   
3  374157065  Flu: The Story of the Great Influenza Pandemic...   
4  393045218                             The Mummies of Urumchi   

            book_author year_of_publication                   publisher  
0    Mark P. O. Morford                2002     Oxford University Press  
1  Richard Bruce Wright                2001       HarperFlamingo Canada  
2          Carlo D'Este                1991             HarperPerennial  
3      Gina Bari Kolata                1999        Farrar Straus Giroux  
4       E. J. W. Barber                1999  W. W. Norton &amp; Company

In [7]:
df = pd.read_csv('E:/Data/BX-Book-Ratings.csv',encoding='latin-1',nrows=10000)

In [8]:
df.head()

user_id        isbn  rating
0   276725  034545104X       0
1   276726   155061224       5
2   276727   446520802       0
3   276729  052165615X       3
4   276729   521795028       6

In [9]:
df.describe()

user_id        rating
count   10000.000000  10000.000000
mean   265844.379600      1.974700
std     56937.189618      3.424884
min         2.000000      0.000000
25%    277478.000000      0.000000
50%    278418.000000      0.000000
75%    278418.000000      4.000000
max    278854.000000     10.000000

In [10]:
df = pd.merge(df,df_books,on='isbn')
df.head()

user_id        isbn  rating            book_title      book_author  \
0   276725  034545104X       0  Flesh Tones: A Novel       M. J. Rose   
1   276726   155061224       5      Rites of Passage       Judith Rae   
2   276727   446520802       0          The Notebook  Nicholas Sparks   
3   278418   446520802       0          The Notebook  Nicholas Sparks   
4   276729  052165615X       3        Help!: Level 1    Philip Prowse   

  year_of_publication                   publisher  
0                2002            Ballantine Books  
1                2001                      Heinle  
2                1996                Warner Books  
3                1996                Warner Books  
4                1999  Cambridge University Press

In [11]:
n_users = df.user_id.nunique()
n_books = df.isbn.nunique()

print('Num. of Users: '+ str(n_users))
print('Num of Books: '+str(n_books))

Num. of Users: 828
Num of Books: 8051


#### Convert ISBN to numeric numbers in order 

In [12]:
# International Standard Book Number (ISBN)
isbn_list = df.isbn.unique()
print(" Length of isbn List:", len(isbn_list))
def get_isbn_numeric_id(isbn):
    #print ("  isbn is:" , isbn)
    itemindex = np.where(isbn_list==isbn)
    return itemindex[0][0]

 Length of isbn List: 8051


#### Do the same for user_id , convert it into numeric and in order

In [13]:
userid_list = df.user_id.unique()
print(" Length of user_id List:", len(userid_list))
def get_user_id_numeric_id(user_id):
    itemindex = np.where(userid_list==user_id)
    return itemindex[0][0]

 Length of user_id List: 828


#### Converting both user_id and isbn to ordered list i.e. from 0...n-1

In [14]:
df['user_id_order'] = df['user_id'].apply(get_user_id_numeric_id)

In [15]:
df['isbn_id'] = df['isbn'].apply(get_isbn_numeric_id)
df.head()

user_id        isbn  rating            book_title      book_author  \
0   276725  034545104X       0  Flesh Tones: A Novel       M. J. Rose   
1   276726   155061224       5      Rites of Passage       Judith Rae   
2   276727   446520802       0          The Notebook  Nicholas Sparks   
3   278418   446520802       0          The Notebook  Nicholas Sparks   
4   276729  052165615X       3        Help!: Level 1    Philip Prowse   

  year_of_publication                   publisher  user_id_order  isbn_id  
0                2002            Ballantine Books              0        0  
1                2001                      Heinle              1        1  
2                1996                Warner Books              2        2  
3                1996                Warner Books              3        2  
4                1999  Cambridge University Press              4        3

#### Re-index columns to build matrix later on

In [16]:
new_col_order = ['user_id_order', 'isbn_id', 'rating', 'book_title', 'book_author','year_of_publication','publisher','isbn','user_id']
df = df.reindex(columns= new_col_order)
df.head()

user_id_order  isbn_id  rating            book_title      book_author  \
0              0        0       0  Flesh Tones: A Novel       M. J. Rose   
1              1        1       5      Rites of Passage       Judith Rae   
2              2        2       0          The Notebook  Nicholas Sparks   
3              3        2       0          The Notebook  Nicholas Sparks   
4              4        3       3        Help!: Level 1    Philip Prowse   

  year_of_publication                   publisher        isbn  user_id  
0                2002            Ballantine Books  034545104X   276725  
1                2001                      Heinle   155061224   276726  
2                1996                Warner Books   446520802   276727  
3                1996                Warner Books   446520802   278418  
4                1999  Cambridge University Press  052165615X   276729

#### Train Test Split

Recommendation Systems  are difficult to evaluate, but you will still learn how to evaluate them. In order to do this, you'll split your data into two sets. However, you won't do your classic X_train,X_test,y_train,y_test split. Instead, you can actually just segement the data into two sets of data:

In [18]:
from sklearn.model_selection import train_test_split
train_data, test_data = train_test_split(df, test_size=0.30)

## Approach: You Will Use Memory-Based Collaborative Filtering

Memory-Based Collaborative Filtering approaches can be divided into two main sections: **user-item filtering** and **item-item filtering**. 

A *user-item filtering* will take a particular user, find users that are similar to that user based on similarity of ratings, and recommend items that those similar users liked. 

In contrast, *item-item filtering* will take an item, find users who liked that item, and find other items that those users or similar users also liked. It takes items as input and outputs other items as recommendations. 

* *Item-Item Collaborative Filtering*: “Users who liked this item also liked …”
* *User-Item Collaborative Filtering*: “Users who are similar to you also liked …”

In both cases, you create a user-book matrix which is built from the entire dataset.

Since you have split the data into testing and training, you will need to create two ``[828 x 8051]`` matrices (all users by all books). This is going to be  a very large matrix

The training matrix contains 70% of the ratings and the testing matrix contains 30% of the ratings.  

In [19]:
#Create two user-book matrices, one for training and another for testing
train_data_matrix = np.zeros((n_users, n_books))
for line in train_data.itertuples():
    train_data_matrix[line[1]-1, line[2]-1] = line[3]  

test_data_matrix = np.zeros((n_users, n_books))
for line in test_data.itertuples():
    test_data_matrix[line[1]-1, line[2]-1] = line[3]

    * You can use the **pairwise_distances** function from sklearn to calculate the cosine similarity. Note, the output will range from 0 to 1 since the ratings are all positive.

In [21]:
from sklearn.metrics.pairwise import pairwise_distances
user_similarity = pairwise_distances(train_data_matrix, metric='cosine')
item_similarity = pairwise_distances(train_data_matrix.T, metric='cosine')

In [22]:
user_similarity

array([[0., 1., 1., ..., 1., 1., 1.],
       [1., 0., 1., ..., 1., 1., 1.],
       [1., 1., 0., ..., 1., 1., 1.],
       ...,
       [1., 1., 1., ..., 0., 1., 1.],
       [1., 1., 1., ..., 1., 0., 1.],
       [1., 1., 1., ..., 1., 1., 0.]])

In [23]:
def predict(ratings, similarity, type='user'):
    if type == 'user':
        mean_user_rating = ratings.mean(axis=1)
        #You use np.newaxis so that mean_user_rating has same format as ratings
        ratings_diff = (ratings - mean_user_rating[:, np.newaxis]) 
        pred = mean_user_rating[:, np.newaxis] + similarity.dot(ratings_diff) / np.array([np.abs(similarity).sum(axis=1)]).T
    elif type == 'item':
        pred = ratings.dot(similarity) / np.array([np.abs(similarity).sum(axis=1)])     
    return pred

In [24]:
item_prediction = predict(train_data_matrix, item_similarity, type='item')
user_prediction = predict(train_data_matrix, user_similarity, type='user')

In [25]:
item_prediction

array([[0.        , 0.00062112, 0.00062112, ..., 0.00062112, 0.00062112,
        0.00062112],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.0578882 , 0.0578882 , 0.0578882 , ..., 0.0578882 , 0.0578882 ,
        0.0578882 ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

### Evaluation
There are many evaluation metrics, but one of the most popular metric used to evaluate accuracy of predicted ratings is *Root Mean Squared Error (RMSE)*.

In [27]:
from sklearn.metrics import mean_squared_error
from math import sqrt
def rmse(prediction, ground_truth):
    prediction = prediction[ground_truth.nonzero()].flatten() 
    ground_truth = ground_truth[ground_truth.nonzero()].flatten()
    return sqrt(mean_squared_error(prediction, ground_truth))

In [28]:
print('User-based CF RMSE: ' + str(rmse(user_prediction, test_data_matrix)))
print('Item-based CF RMSE: ' + str(rmse(item_prediction, test_data_matrix)))

User-based CF RMSE: 7.755063277047862
Item-based CF RMSE: 7.754569972149767


In [29]:
df

user_id_order  isbn_id  rating                book_title  \
0                 0        0       0      Flesh Tones: A Novel   
1                 1        1       5          Rites of Passage   
2                 2        2       0              The Notebook   
3                 3        2       0              The Notebook   
4                 4        3       3            Help!: Level 1   
...             ...      ...     ...                       ...   
8696             96     8046       7        A Map of the World   
8697             96     8047       0    The Accidental Tourist   
8698             96     8048       0     If Morning Ever Comes   
8699             96     8049       9        Unnatural Exposure   
8700             96     8050       0  Only Love (Magical Love)   

                    book_author year_of_publication  \
0                    M. J. Rose                2002   
1                    Judith Rae                2001   
2               Nicholas Sparks                1996   
3               Nicholas Sparks                1996   
4                 Philip Prowse                1999   
...                         ...                 ...   
8696              Jane Hamilton                1999   
8697                 Anne Tyler                1994   
8698                 Anne Tyler                1983   
8699  Patricia Daniels Cornwell                1998   
8700                Erich Segal                1998   

                       publisher        isbn  user_id  
0               Ballantine Books  034545104X   276725  
1                         Heinle   155061224   276726  
2                   Warner Books   446520802   276727  
3                   Warner Books   446520802   278418  
4     Cambridge University Press  052165615X   276729  
...                          ...         ...      ...  
8696      Anchor Books/Doubleday   385720106      243  
8697    Berkley Publishing Group   425092917      243  
8698    Berkley Publishing Group   425098834      243  
8699    Berkley Publishing Group   425163407      243  
8700    Berkley Publishing Group   425164403      243  

[8701 rows x 9 columns]

In [31]:
book_pivot = df.pivot_table(columns='user_id', index='book_title', values="rating")
book_pivot.fillna(0, inplace=True)

In [34]:
book_pivot.shape
book_pivot.iloc[:5,:10]

user_id                                              2    8    9    10   12  \
book_title                                                                    
 Little Comic Shop of Horrors (Give Yourself Go...  0.0  0.0  0.0  0.0  0.0   
007 El Mundo Nunca Es Suficiente                    0.0  0.0  0.0  0.0  0.0   
01-01-00: A Novel of the Millennium                 0.0  0.0  0.0  0.0  0.0   
01-01-00: The Novel of the Millennium               0.0  0.0  0.0  0.0  0.0   
100 Best-Loved Poems (Dover Thrift Editions)        0.0  0.0  0.0  0.0  0.0   

user_id                                              14   16   17   19   20  
book_title                                                                   
 Little Comic Shop of Horrors (Give Yourself Go...  0.0  0.0  0.0  0.0  0.0  
007 El Mundo Nunca Es Suficiente                    0.0  0.0  0.0  0.0  0.0  
01-01-00: A Novel of the Millennium                 0.0  0.0  0.0  0.0  0.0  
01-01-00: The Novel of the Millennium               0.0  0.0  0.0  0.0  0.0  
100 Best-Loved Poems (Dover Thrift Editions)        0.0  0.0  0.0  0.0  0.0

In [36]:
from scipy.sparse import csr_matrix
book_sparse = csr_matrix(book_pivot)
book_sparse

<7850x828 sparse matrix of type '<class 'numpy.float64'>'
	with 2235 stored elements in Compressed Sparse Row format>

In [37]:
from sklearn.neighbors import NearestNeighbors
model = NearestNeighbors(algorithm='brute')
model.fit(book_sparse)

NearestNeighbors(algorithm='brute')

In [50]:
n = np.random.choice(book_pivot.shape[0])
n

4415

In [59]:
distances, suggestions = model.kneighbors(book_pivot.iloc[n, :].values.reshape(1, -1))

In [60]:
distances.shape, suggestions.shape

((1, 5), (1, 5))

In [61]:
pd.DataFrame(book_pivot.iloc[n, :]).columns

Index(['On Writing Well 5ED'], dtype='object')

In [62]:
for i in range(5):
    print(book_pivot.index[suggestions[0,i]])

Quality Is Free: The Art of Making Quality Certain
Q Clearance
Quantum Leap: Independence : A Novel (Quantum Leap)
Quantum Leap: Obsessions (Quantum Leap)
Quanah Parker (War Chiefs No. 4)
